sql,opt: implement region-column inference

DrewKimball opened the pull request:
#### sql,builtins: add nil-check for functions that use evalCtx.Regions

The optimizer tests do not initialize the `evalCtx.Regions` field.
This could previously cause NPE when the optimizer attempted constant
folding for a function that uses the field. This commit hardens all
such builtin functions against NPE.

Epic: None

Release note: None

#### opt: add support for region column inference to catalog and test catalog

This commit adds support for the new storage param
`infer_rbr_region_col_using_constraint` to the optimizer and test catalogs.

Informs #148243

Release note: None

#### opt: implement region-column inference

This commit finishes the implementation for the new storage parameter
`infer_rbr_region_col_using_constraint`. If a foreign key constraint
is specified via that setting, the constraint will be used to plan
a lookup join into the parent table to retrieve the correct values for
the region column on INSERT, UPDATE, and UPSERT operations. This will
make it easier for customers to add the `crdb_region` column to their
foreign-key constraints, since they no longer have to guarantee that
an insert into the child table happens in the same region as the
matching parent row.

Fixes #148243

Release note (sql change): Added support for automatically determining
the region column for a REGIONAL BY ROW table using a foreign key constraint.
The foreign key is specified by setting a new table storage parameter
`infer_rbr_region_col_using_constraint`, and must contain the region
column. This can be useful for applications that are unable to guarantee
that a child row is inserted or updated from the same region as the
matching parent row.

DrewKimball committed:
opt: add support for region column inference to catalog and test catalog

This commit adds support for the new storage param
`infer_rbr_region_col_using_constraint` to the optimizer and test catalogs.

Informs #148243

Release note: None

DrewKimball committed:
opt: implement region-column inference

This commit finishes the implementation for the new storage parameter
`infer_rbr_region_col_using_constraint`. If a foreign key constraint
is specified via that setting, the constraint will be used to plan
a lookup join into the parent table to retrieve the correct values for
the region column on INSERT, UPDATE, and UPSERT operations. This will
make it easier for customers to add the `crdb_region` column to their
foreign-key constraints, since they no longer have to guarantee that
an insert into the child table happens in the same region as the
matching parent row.

Fixes #148243

Release note (sql change): Added support for automatically determining
the region column for a REGIONAL BY ROW table using a foreign key constraint.
The foreign key is specified by setting a new table storage parameter
`infer_rbr_region_col_using_constraint`, and must contain the region
column. This can be useful for applications that are unable to guarantee
that a child row is inserted or updated from the same region as the
matching parent row.

DrewKimball commented:
The first two commits are #148539.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148540)
<!-- Reviewable:end -->

michae2 reviewed (commented):
Nice job putting this together!

I have some design-y questions which maybe are too late in the process. You can push back against them. Maybe these have already been discussed between you and @dikshant?

> Reviewed 1 of 1 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-:-OThVu497nbggvuIq9LG:b-m1tjev)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 432 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThUudN-3kKAuMZH8xb:-OThUudN-3kKAuMZH8xc:b-5o0vqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L432)):*
> ```
>   k INT,
>   v INT,
>   CONSTRAINT foo FOREIGN KEY (crdb_region, k) REFERENCES parent_rbt (region, k)
> ```

I'm a little sad this requires explicitly including the crdb_region column in the FK. The beauty of a hidden crdb_region column is that a non-RBR table can be converted directly into an RBR table with simply:

```sql
ALTER TABLE mytable SET LOCALITY REGIONAL BY ROW;
```

It would be really cool if we could simply write:

```sql
ALTER TABLE mytable SET (infer_rbr_region_col_using_constraint = 'foo') LOCALITY REGIONAL BY ROW;
```

And that FK constraint would get an implicit region column added, without having to manually alter the FK separately.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 955 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThGVIlD4tAQKXvORhH:-OThGVIlD4tAQKXvORhI:b4459e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L955)):*
> ```Go
> 			"lookup columns using constraint %q must be a lax key", lookupFK.Name()))
> 	}
> 	joinPrivate := &memo.JoinPrivate{Flags: memo.PreferLookupJoinIntoRight}
> ```

Should we put the `PreferLookupJoinIntoRight` flag behind `mb.b.evalCtx.SessionData().PreferLookupJoinsForFK` to match FK checks?

Does this always turn into a LOJ? Looks like it often does. I'll read through more of the testcases.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 980 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThIH8kCK28MfQaTVQc:-OThIH8kCK28MfQaTVQd:b-tptknu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L980)):*
> ```Go
> 
> 	// Whether a row is inserted or updated, it will use the newly calculated
> 	// value for the region column.
> ```

Will this always override an explicit assignment to the region column? If so, that seems a little surprising.

Maybe a similar question, how does this interact with DEFAULT for the region column? In a way, it seems like this is a new kind of DEFAULT for the column. I could even think about it as a subquery in the DEFAULT clause, something like:

```sql
CREATE TABLE parent (
  p INT PRIMARY KEY
) LOCALITY REGIONAL BY ROW;

CREATE TABLE child (
  region crdb_internal_region NOT NULL DEFAULT (SELECT CASE WHEN crdb_region IS NULL THEN gateway_region() ELSE crdb_region END FROM parent WHERE parent.p = child.p),
  c INT PRIMARY KEY,
  p INT NOT NULL,
  CONSTRAINT foo FOREIGN KEY (region, p) REFERENCES parent (crdb_region, p) ON UPDATE CASCADE ON DELETE CASCADE,
) WITH (infer_rbr_region_col_using_constraint = 'foo') LOCALITY REGIONAL BY ROW AS region;
```

I'm not suggesting we change the syntax or anything like that, just want to make sure we're thinking about how this interacts with DEFAULT and explicit assignments.

One (unsolicited) idea: maybe whenever the `infer_rbr_region_col_using_constraint` storage parameter is set, we should change the DEFAULT for the region column to be some new builtin function like `infer_rbr_region_col_using_constraint()` and then we should only add the lookup join when this builtin is used in a projection. That way explicit assignments to the region column won't produce the lookup join.
___
*[`pkg/sql/opt_catalog.go` line 1577 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OTddxdyCtKO365MSTxN:-OTdZq7qEcTuqBg4ZnMg:bwyqhbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/97b9d28f7b592db8d0656b54ba5683e78f6b6cf6/pkg/sql/opt_catalog.go#L1577)):*
> ```Go
> // RegionalByRowUsingConstraint is part of the cat.Table interface.
> func (ot *optTable) RegionalByRowUsingConstraint() cat.ForeignKeyConstraint {
> 	if id := ot.desc.GetRegionalByRowUsingConstraint(); id != catid.ConstraintID(0) {
> ```

Should this start by checking whether the table is RBR?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @DrewKimball, and @mgartner)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 432 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThUudN-3kKAuMZH8xb:-OTi3wBVDyNEJbMd3WIe:bx1kh9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L432)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm a little sad this requires explicitly including the crdb_region column in the FK. The beauty of a hidden crdb_region column is that a non-RBR table can be converted directly into an RBR table with simply:

```sql
ALTER TABLE mytable SET LOCALITY REGIONAL BY ROW;
```

It would be really cool if we could simply write:

```sql
ALTER TABLE mytable SET (infer_rbr_region_col_using_constraint = 'foo') LOCALITY REGIONAL BY ROW;
```

And that FK constraint would get an implicit region column added, without having to manually alter the FK separately.
</blockquote></details>

@dikshant and I talked about this just now, and he pointed out that we can always add support for implicit region in FK constraints later. It's a superset of this behavior. So no need to change this for 25.3.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @DrewKimball, and @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 980 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThIH8kCK28MfQaTVQc:-OTi4kAgCsSvymXJc5Pn:bmar4gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L980)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Will this always override an explicit assignment to the region column? If so, that seems a little surprising.

Maybe a similar question, how does this interact with DEFAULT for the region column? In a way, it seems like this is a new kind of DEFAULT for the column. I could even think about it as a subquery in the DEFAULT clause, something like:

```sql
CREATE TABLE parent (
  p INT PRIMARY KEY
) LOCALITY REGIONAL BY ROW;

CREATE TABLE child (
  region crdb_internal_region NOT NULL DEFAULT (SELECT CASE WHEN crdb_region IS NULL THEN gateway_region() ELSE crdb_region END FROM parent WHERE parent.p = child.p),
  c INT PRIMARY KEY,
  p INT NOT NULL,
  CONSTRAINT foo FOREIGN KEY (region, p) REFERENCES parent (crdb_region, p) ON UPDATE CASCADE ON DELETE CASCADE,
) WITH (infer_rbr_region_col_using_constraint = 'foo') LOCALITY REGIONAL BY ROW AS region;
```

I'm not suggesting we change the syntax or anything like that, just want to make sure we're thinking about how this interacts with DEFAULT and explicit assignments.

One (unsolicited) idea: maybe whenever the `infer_rbr_region_col_using_constraint` storage parameter is set, we should change the DEFAULT for the region column to be some new builtin function like `infer_rbr_region_col_using_constraint()` and then we should only add the lookup join when this builtin is used in a projection. That way explicit assignments to the region column won't produce the lookup join.
</blockquote></details>

(A TODO and a GH issue about this would also be fine with me.)


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @DrewKimball, and @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 980 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThIH8kCK28MfQaTVQc:-OUCa4Ih2D5Jn20FrFjW:b-l3y1pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L980)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

(A TODO and a GH issue about this would also be fine with me.)
</blockquote></details>

We talked about this a few days ago. You pointed out that DEFAULT expressions, computed columns, user-provided expressions, and BEFORE triggers all have to be evaluated before we join to the parent table in order to have the values for the other columns in the FK. That makes sense.

But I can't shake the feeling that this behavior is very surprising. I think it will appear that the database is operating incorrectly to someone explicitly assigning a region.

How about a compromise? Hypothetically what would you think of something like this:

First, change the projection on top of the join from:

```
CASE
  WHEN "lookupRegionColID" IS NULL THEN "originRegionColID"
  ELSE "lookupRegionColID"
END
```

to:

```
CASE
  WHEN "originRegionColID" IS NOT NULL THEN "originRegionColID"
  WHEN "lookupRegionColID" IS NOT NULL THEN "lookupRegionColID"
  ELSE default_to_database_primary_region(gateway_region())
END
```

then as part of setting `infer_rbr_region_col_using_constraint` set the DEFAULT to NULL (or require to user to do so).

This way, explicit assignments will take precedence.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt_catalog.go` line 1577 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OTddxdyCtKO365MSTxN:-OV9fPXN49VC6Z0U5m9q:b-7y9mlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/97b9d28f7b592db8d0656b54ba5683e78f6b6cf6/pkg/sql/opt_catalog.go#L1577)):*
>Should this start by checking whether the table is RBR?

It's not strictly necessary, but it'll be more clear if we do. Done.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 955 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThGVIlD4tAQKXvORhH:-OV9fkkh8LUdDGR37TJ2:bujxrt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L955)):*
>Should we put the `PreferLookupJoinIntoRight` flag behind `mb.b.evalCtx.SessionData().PreferLookupJoinsForFK` to match FK checks?

Done.

>Does this always turn into a LOJ? Looks like it often does. I'll read through more of the testcases.

A lookup join should be possible whenever there's a suitable index, which I think will be the case unless someone is using `UNIQUE WITHOUT INDEX`. It should be a LOJ whenever the parent is a RBR table - if not, that's likely coster problem, since the region column partitions every index in an RBR table.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 980 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OThIH8kCK28MfQaTVQc:-OV9hI1W0Acxnf0RqmCZ:b-puqsu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5496d6921a3a380cd7aa14ef7bcf8fc0afa8470/pkg/sql/opt/optbuilder/mutation_builder.go#L980)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

We talked about this a few days ago. You pointed out that DEFAULT expressions, computed columns, user-provided expressions, and BEFORE triggers all have to be evaluated before we join to the parent table in order to have the values for the other columns in the FK. That makes sense.

But I can't shake the feeling that this behavior is very surprising. I think it will appear that the database is operating incorrectly to someone explicitly assigning a region.

How about a compromise? Hypothetically what would you think of something like this:

First, change the projection on top of the join from:

```
CASE
  WHEN "lookupRegionColID" IS NULL THEN "originRegionColID"
  ELSE "lookupRegionColID"
END
```

to:

```
CASE
  WHEN "originRegionColID" IS NOT NULL THEN "originRegionColID"
  WHEN "lookupRegionColID" IS NOT NULL THEN "lookupRegionColID"
  ELSE default_to_database_primary_region(gateway_region())
END
```

then as part of setting `infer_rbr_region_col_using_constraint` set the DEFAULT to NULL (or require to user to do so).

This way, explicit assignments will take precedence.
</blockquote></details>

The idea to let explicit assignments override this feature makes sense to me. I think we can simplify things even further - if the region column was explicitly specified, skip the join. That does mean that a user could specify a nonexistent value or NULL and get an error, but this seems fine. I've updated the behavior and tests accordingly.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 5 of 14 files at r1, 45 of 45 files at r6, 23 of 23 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @mw5h)

___
*[`pkg/sql/alter_table_locality.go` line 363 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE40YI7lZ18lSuHS-d:-OVE40YI7lZ18lSuHS-e:b-5qsijx) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/alter_table_locality.go#L363)):*
> ```Go
> 
> 	// Disallow changing the region column if the table is using a FK constraint
> 	// to determine values for the region column.
> ```

good thinking!
___
*[`pkg/sql/create_table.go` line 2464 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE4FgoB1HGHfOi2_Zv:-OVE4FgoB1HGHfOi2_Zw:b-ga8cl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/create_table.go#L2464)):*
> ```Go
> 		return nil, pgerror.Newf(
> 			pgcode.InvalidParameterValue,
> 			`storage parameter "%s" can only be set on regional by row tables`,
> ```

nit: capitalize REGIONAL BY ROW?
___
*[`pkg/sql/catalog/tabledesc/structured.go` line 2585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE2let00sQZr1pLXFg:-OVE2let00sQZr1pLXFh:bd1j79v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3812a706aadb205720ab0646b65a9190987bd3d/pkg/sql/catalog/tabledesc/structured.go#L2585)):*
> ```Go
> 	}
> 	if usingFK := desc.GetRegionalByRowUsingConstraint(); usingFK != descpb.ConstraintID(0) {
> 		if constraint := catalog.FindConstraintByID(desc, usingFK); constraint != nil {
> ```

when would `constraint == nil` here? should that be an assertion error? this function shouldn't panic and it doesn't currently return an error, so that could be a complicated change, huh?
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2367 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE58sb7FTGzh9SSOyH:-OVE58sb7FTGzh9SSOyI:b-2kiyuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/catalog/tabledesc/validate.go#L2367)):*
> ```Go
> 	// needed because the values of every (possibly computed) foreign-key column
> 	// must be known in order to determine the value for the region column.
> 	for _, col := range tableDesc.NonDropColumns() {
> ```

question (maybe for SQL Foundations): is ok to ignore drop columns in this case?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @mw5h)

___
*[`pkg/sql/alter_table_locality.go` line 363 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE40YI7lZ18lSuHS-d:-OVF7PSZ2HH4srRevOAf:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/alter_table_locality.go#L363)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

good thinking!
</blockquote></details>

Thanks!
___
*[`pkg/sql/create_table.go` line 2464 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE4FgoB1HGHfOi2_Zv:-OVF7p3j4r9jO0iEocvg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/create_table.go#L2464)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: capitalize REGIONAL BY ROW?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/tabledesc/structured.go` line 2585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE2let00sQZr1pLXFg:-OVF8F7KB1K_dZWOFkYS:b-pbhfvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3812a706aadb205720ab0646b65a9190987bd3d/pkg/sql/catalog/tabledesc/structured.go#L2585)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

when would `constraint == nil` here? should that be an assertion error? this function shouldn't panic and it doesn't currently return an error, so that could be a complicated change, huh?
</blockquote></details>

Good point. It should never happen, particularly since we check this when validating the descriptor, but it turned out to be pretty easy to just plumb the error.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2367 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVE58sb7FTGzh9SSOyH:-OVF9aAm4-FlcO6yrOev:bazu3mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/aec2f9bac8a1bc77075ab20986066435572ba4c2/pkg/sql/catalog/tabledesc/validate.go#L2367)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

question (maybe for SQL Foundations): is ok to ignore drop columns in this case?
</blockquote></details>

This is based off of `ValidateColumnHasNoDependents`, which also uses `NonDropColumns()`.


<!-- Sent from Reviewable.io -->

mw5h commented on pkg/sql/opt/optbuilder/mutation_builder.go:
```diff
@@ -861,6 +883,138 @@ func (mb *mutationBuilder) addSynthesizedComputedCols(colIDs opt.OptionalColList
+			includeInverted:  false,
+		}),
+		&tree.IndexFlags{IgnoreForeignKeys: true},
+		noRowLocking,
```
You might need locking for non-serializable isolations.

yuzefovich replied:
Also maybe for serializable too when `ImplicitFKLockingForSerializable` is `true` (if following `fkCheckHelper.buildOtherTableScan` as the example).
<!-- Reviewable comment -OVFt6jm5yUizTS2goI0:b-d6hv6p -->
<!-- Sent from Reviewable.io -->

DrewKimball replied:
Good point. It's not necessary for correctness, but seems desirable for the user experience, since otherwise the FK check could fail after the mutation. I added a read-committed test.
<!-- Reviewable comment -OVGS9Q18Hd4hG5_w2dd:b-1jqhl5 -->
<!-- Sent from Reviewable.io -->

mw5h commented on pkg/sql/opt/optbuilder/mutation_builder.go:
```diff
@@ -861,6 +883,138 @@ func (mb *mutationBuilder) addSynthesizedComputedCols(colIDs opt.OptionalColList
+		mb.b.allocScope(),
+		true, /* disableNotVisibleIndex */
+		// The scan is exempt from RLS to maintain data integrity.
+		cat.PolicyScopeExempt,
```
Is this okay? Could an attacker use this to probe for table data they don't have permission to view? It feels like this could be used to ask questions like "is this value present in this other table?"

DrewKimball replied:
If it's not OK, then it seems like FK checks would also not be ok. Like this:
```
CREATE TABLE parent (x INT, y INT, PRIMARY KEY (x, y))
-- Some RLS constraint saying the user can only read rows from "parent" where "y" > 0

CREATE TABLE child (
  x INT,
  y INT,
  FOREIGN KEY (x, y) REFERENCES parent (x, y)
)

-- If the insert succeeds, we know there's a row for which parent.y = -1.
INSERT INTO child VALUES (1, -1);
```
I think defining these sorts of constraints implies that the user is OK with information leaking through this channel.
<!-- Reviewable comment -OVGLTRbAqpUCZy89GLA:bbokvi -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Great work!

> Reviewed 8 of 8 files at r8, 6 of 6 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-:-OVFy_ds7zr0HHJebwXL:b-9dwnzc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, @michae2, and @mw5h)

___
*[`pkg/sql/opt/testutils/testcat/create_table.go` line 424 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFlxKqDKcgzNZ_xp7i:-OVFlxKqDKcgzNZ_xp7j:b5woqf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9aa5b2e116c81e24e51e3e9a921e6e2e84514bd/pkg/sql/opt/testutils/testcat/create_table.go#L424)):*
> ```Go
> 	}
> 
> 	// Allow specifying the name of a FK constraint to use for lookup up region
> ```

nit: "lookup up".
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 924 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFsBAi7pZX5JhDhUbU:-OVFsBAj-EFrZscbdDAV:b-x5i69x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/sql/opt/optbuilder/mutation_builder.go#L924)):*
> ```Go
> 	// Resolve the referenced table.
> 	refTabDescID := int64(lookupFK.ReferencedTableID())
> 	refTab := mb.b.resolveTableRef(&tree.TableRef{TableID: refTabDescID}, privilege.SELECT)
> ```

Do we need to check the SELECT privileges on the referenced table? If not, consider leaving a comment why.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 967 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFrU0p6XHP2EhaXyrH:-OVFrU0p6XHP2EhaXyrI:bkdzme7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/sql/opt/optbuilder/mutation_builder.go#L967)):*
> ```Go
> 			includeInverted:  false,
> 		}),
> 		&tree.IndexFlags{IgnoreForeignKeys: true},
> ```

nit: should we set `AvoidFullScan` if the corresponding session variable is set, for completeness?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 478 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFyIEa8OqFKsCP7WdV:-OVFyIEa8OqFKsCP7WdW:b-wb32bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L478)):*
> ```
> subtest end
> 
> # ==============================================================================
> ```

Nice tests! Do you think it'd be worth adding a test case where the mutations for which region inference happens is within a routine? (Up to you - it's the only scenario I didn't see covered.)
___
*[`pkg/sql/sem/builtins/builtins.go` line 7575 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFmxUR86BnU5DCqWed:-OVFmxUR86BnU5DCqWee:bi72aq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9aa5b2e116c81e24e51e3e9a921e6e2e84514bd/pkg/sql/sem/builtins/builtins.go#L7575)):*
> ```Go
> 			func(ctx context.Context, evalCtx *eval.Context, s string) (tree.Datum, error) {
> 				var regionConfig eval.DatabaseRegionConfig
> 				if evalCtx.Regions != nil {
> ```

nit: there are 4 other builtins that access `evalCtx.Regions`. It feels like adding a nil check should be extracted into a separate commit with all builtins adjusted.

Also why we did need this fix? Is it because we have incomplete eval context in the optbuilder tests?


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql,builtins: add nil-check for functions that use evalCtx.Regions

The optimizer tests do not initialize the `evalCtx.Regions` field.
This could previously cause NPE when the optimizer attempted constant
folding for a function that uses the field. This commit hardens all
such builtin functions against NPE.

Epic: None

Release note: None

DrewKimball reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-:-OVGdXEy2AzfcGrEOkNn:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @mw5h, and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 924 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFsBAi7pZX5JhDhUbU:-OVGHDYcDQ9Gi6kmYaeD:b7d9lh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/sql/opt/optbuilder/mutation_builder.go#L924)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Do we need to check the SELECT privileges on the referenced table? If not, consider leaving a comment why.
</blockquote></details>

We do check for SELECT privileges in `resolveTableRef`, and I think it makes sense to do so since we check the same for FK checks and cascades.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 967 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFrU0p6XHP2EhaXyrH:-OVGHYQg33qqq-lmmbD1:b-y0hhb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/sql/opt/optbuilder/mutation_builder.go#L967)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we set `AvoidFullScan` if the corresponding session variable is set, for completeness?
</blockquote></details>

Nice catch, Done.
___
*[`pkg/sql/opt/testutils/testcat/create_table.go` line 424 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFlxKqDKcgzNZ_xp7i:-OVGH0MuD2hbSfjjXPdr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9aa5b2e116c81e24e51e3e9a921e6e2e84514bd/pkg/sql/opt/testutils/testcat/create_table.go#L424)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: "lookup up".
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7575 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFmxUR86BnU5DCqWed:-OVG_od3-I3jcgi_AIeM:b-v8om09) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9aa5b2e116c81e24e51e3e9a921e6e2e84514bd/pkg/sql/sem/builtins/builtins.go#L7575)):*
>nit: there are 4 other builtins that access `evalCtx.Regions`. It feels like adding a nil check should be extracted into a separate commit with all builtins adjusted.

Done.

>Also why we did need this fix? Is it because we have incomplete eval context in the optbuilder tests?

Yes, the optimizer was trying to constant-fold the region default function and hit a NPE.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 478 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFyIEa8OqFKsCP7WdV:-OVGYfHK1Qk2C0T5X-E1:bwq7l89) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L478)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Nice tests! Do you think it'd be worth adding a test case where the mutations for which region inference happens is within a routine? (Up to you - it's the only scenario I didn't see covered.)
</blockquote></details>

Sure, I think that's a good idea. Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

nit: update the PR description to include all 3 commits

> Reviewed 1 of 6 files at r9, 8 of 51 files at r10, 46 of 46 files at r11, 44 of 44 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-:-OVIirzL-IRPxNmOP6L1:b-2ycpa9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2, @mw5h, and @yuzefovich)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 478 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVFyIEa8OqFKsCP7WdV:-OVIpZ8A6XVIc-8jyMXu:bwaakwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ce44fff3b5ef307a2fa6b869d02999a87e4c1d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L478)):*
> Nice tests!

:heavy_plus_sign: Nice job enumerating a lot of interesting feature combinations!
___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 47 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIpAbp0KmfzxOHevCg:-OVIpAbp0KmfzxOHevCh:byd6ovk) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L47)):*
> ```
>   v INT,
>   region crdb_internal_region,
>   UNIQUE WITHOUT INDEX (k, region)
> ```

What's the reason for this `UNIQUE WITHOUT INDEX` on a global table?
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 942 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIkhzE5w-4fA92KNjt:-OVIkhzE5w-4fA92KNju:b-tq2i1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L942)):*
> ```Go
> 			originRegionColID = originColID
> 			lookupRegionColID = refColID
> 			continue
> ```

nit: might be good to add a little explanation here or in the comment above on in the comment at the top of the function to point out that we're looking up on all columns except the region, 
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 963 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIl7-54569P0o_S09W:-OVIl7-54569P0o_S09X:b-vkkhzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L963)):*
> ```Go
> 	// when reading from the parent table. This prevents concurrent transactions
> 	// from invalidating the looked-up region column value. This isn't necessary
> 	// for correctness since FK checks still run, but avoids
> ```

nit: unfinished comment
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1008 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIlRjv5Nuu4Z9m1Yq8:-OVIlRjv5Nuu4Z9m1Yq9:b-5wko2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L1008)):*
> ```Go
> 	)
> 	regionColType := md.ColumnMeta(originRegionColID).Type
> 	caseExpr := mb.b.factory.ConstructCase(
> ```

nit: an explanation of this CASE expression would be helpful
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 106 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIn4rg61O69RV032_r:-OVIn4rg61O69RV032_s:b-ehpa8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L106)):*
> ```
>  ├── project
>  │    ├── project
>  │    │    ├── left-join (lookup parent@parent_crdb_region_a_b_key)
> ```

is there a way to show the lookup expr without making the output too verbose? it seems like a fairly important thing to check in these tests
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 301 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVInnHZDkePcC4tywbs:-OVInnHZDkePcC4tywbt:blj35ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L301)):*
> ```
> # Case with a computed column in the FK.
> opt locality=(region=east) format=(hide-all)
> INSERT INTO child_computed (x, z) VALUES (1, 2);
> ```

super nit: use something other than `2` for `z` so it has a different value than the computed `y`, making the plan a bit easier to follow 
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 344 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVInwvH1GZXJJMwXV7o:-OVInwvH1GZXJJMwXV7p:b-qqbu57) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L344)):*
> ```
> 
> opt locality=(region=east) format=(hide-all)
> UPSERT INTO child_computed (x, z) VALUES (1, 2);
> ```

nit: ditto - use something other than 2


<!-- Sent from Reviewable.io -->

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks!

DrewKimball reviewed (commented):
Thanks for the thorough reviews!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-:-OVJ9TIxCMTdDlrWqHCG:b-u9bhow)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, @mw5h, and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 942 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIkhzE5w-4fA92KNjt:-OVJ5ioS-L6ylZGb-ndP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L942)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: might be good to add a little explanation here or in the comment above on in the comment at the top of the function to point out that we're looking up on all columns except the region,
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 963 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIl7-54569P0o_S09W:-OVJ6fbz1uNUG82QrEWI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L963)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: unfinished comment
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1008 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIlRjv5Nuu4Z9m1Yq8:-OVJ6udOBUSJecV_fFv3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/mutation_builder.go#L1008)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: an explanation of this CASE expression would be helpful
</blockquote></details>

Done.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key` line 47 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIpAbp0KmfzxOHevCg:-OVJ7g7x6pl9YIv-QFjL:bnt00mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key#L47)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What's the reason for this `UNIQUE WITHOUT INDEX` on a global table?
</blockquote></details>

It technically doesn't add any information, but foreign keys require a constraint on all the foreign key columns; see https://github.com/cockroachdb/cockroach/issues/64619. I wonder if we should reopen that issue - seems like an unfortunate rough edge in our MR syntax considering we recommend adding the region column to FKs.
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 106 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVIn4rg61O69RV032_r:-OVJ4uKZ0MlgBpVuc3J8:b-45t6bv) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L106)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

is there a way to show the lookup expr without making the output too verbose? it seems like a fairly important thing to check in these tests
</blockquote></details>

If I add `show-columns` the lookup expr will be displayed. Done.
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 301 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVInnHZDkePcC4tywbs:-OVJ57IiBh0gsg-yjziT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L301)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

super nit: use something other than `2` for `z` so it has a different value than the computed `y`, making the plan a bit easier to follow
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/testdata/region_lookup` line 344 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/148540#-OVInwvH1GZXJJMwXV7o:-OVJ5GmOAiO7S4QYgClr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da41f08bb7aa9ffa124e57ed87becdd3ccc395f9/pkg/sql/opt/optbuilder/testdata/region_lookup#L344)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: ditto - use something other than 2
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r=michae2,mw5h,mgartner,yuzefovich
